common/meta/types/src/share.rs

```diff
@@ -12,11 +12,29 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+// both id and name will not change after created
+// id used to distinguish share with same name (but never the same time)
 #[derive(serde::Serialize, serde::Deserialize, Clone, Debug, Eq, PartialEq, Default)]
 pub struct ShareInfo {
-    share_name: String,
-    database: Option<String>,
-    tables: Vec<String>,
+    pub id: u64,
+    pub name: String,
+    pub meta: ShareMeta,
+}
```

Comment:
Is an `id` enough to identify a `ShareMeta`?

As in `TableInfo`, a tuple of `id, ver`(`TableIdent`) is required to uniquely identify a `TableMeta`. Because a `TableMeta` can be updated and the `TableIdent.ver` will be increased.
https://github.com/datafuselabs/databend/blob/fbba0fca848662eb1e9044d71fc2caf33e38c948/common/meta/types/src/table.rs#L82-L99

Reply:
I'm not sure: table version refer to things like snapshot?  need it because a query/transaction  need to refer the same version of data.

for shareInfo, user just use the newest updated share  to find the available object inside (a list of ref to tables)
so I think it is enough

Reply:
@drmingdrmer share is at the database level, maybe has similar conditions, but not very clear to me now.
I also saw SEQ_DATABASE_META_ID, but not found it usage.

currently only think of it as a role object,  a white list of the database/tables behind it.

Reply:
Anything that supports modification should have a `version` to identify a specific version of it, inside metasrv.

For a client outside metasrv, without a `version`, one update operation may override another update operation, which may be not expected.

Reply:
Anything that supports modification should have a `version` to identify a specific version of it, inside metasrv.

For a client outside metasrv, without a `version`, one update operation may override another update operation, which may be not expected.

Reply:
got it,  the `version`  just play the role of seqV.seq in KV api. thank you!

Reply:
> For a client outside metasrv, without a version, one update operation may override another update operation, which may be not expected.

does unexpected override happen even if all update is done in txn.

Reply:
> got it, the `version` just play the role of seqV.seq in KV api. thank you!

Right. They are the same thing with different names :(



> does unexpected override happen even if all update is done in txn.

Yes. an update request may not specify an expected version to update, i.e., it tries to override existing value without considering the version, by using the argument `MatchSeq::Any`.

```rust
/// Describes what `seq` an operation must match to take effect.
/// Every value written to meta data has a unique `seq` bound.
/// Any conditioned or non-conditioned write operation can be done through the corresponding MatchSeq.
#[derive(Serialize, Deserialize, Debug, Copy, Clone, PartialEq, Eq)]
pub enum MatchSeq {
    /// Any value is acceptable, i.e. does not check seq at all.
    Any,

    /// To match an exact value of seq.
    /// E.g., CAS updates the exact version of some value,
    /// and put-if-absent adds a value only when seq is 0.
    Exact(u64),

    /// To match a seq that is greater-or-equal some value.
    /// E.g., GE(1) perform an update on any existent value.
    GE(u64),
}
```

Reply:
> what if the value is never update as a whole, instead api finally call share info.add_table(tablename) in txn , the func check tablename existence
> 
> in this case, no need for seq?

If all operations on a `share` are [commutative](https://en.wikipedia.org/wiki/Commutative_property), no `seq` is required. 

The simplest example of a commutative value is an integer, and only accept addition and substaction operation: `x+1+3` and `x+3+1` result in the same value.

An integer that accepts `+-*/` is not commutative: `x+1*3` may not equal `x*3+1`.

A set of integer that provides methods `Set::insert(i)` and `Set::remove(i)` is not commutative: `{0,1,2}.insert(1).remove(1)` does not equal `{0,1,2}.remove(1).insert(1)`.

A non-commutative value needs a `seq` to serialize every operation on it.

> another choice is make share immutalbe store sub tables in another keyspace

It sounds like a **pointer**, right? Unfortunately, updating a pointer is not a commutative operation. `ptr = a; ptr = b` and `ptr = b; ptr = a` will result in different values.

Reply:
If you can be sure that executing any set of operations in any permutation always results in a valid state, e.g., no lost-update issue, etc, a `seq` is not required.